Common/Tests/Utilities.UI/UI/ReplWindowProxy.cs

```diff
@@ -550,40 +479,52 @@ public class SessionHolder<T> : IDisposable where T : IIntellisenseSession {
+                });
+            }
+#if DEV14_OR_LATER
+            bool canExecute = Invoke(() => _window.Operations.TrySubmitStandardInput() || _window.Operations.Return());
+#else
             bool canExecute = Invoke(() => {
-                if (!string.IsNullOrEmpty(line)) {
-                _editorOperations.InsertText(line);
-                    }
+                var rw = (ReplWindow)_window;
                 var res = _window.Evaluator.CanExecuteText(_window.CurrentLanguageBuffer.CurrentSnapshot.GetText());
-                var pkgCmdSet = VSConstants.VSStd2K;
-                ErrorHandler.ThrowOnFailure(_window.Exec(
-                    ref pkgCmdSet,
-                    (uint)VSConstants.VSStd2KCmdID.RETURN,
-                    0,
-                    IntPtr.Zero,
-                    IntPtr.Zero
-                ));
+                if (res && rw.CaretInStandardInputRegion) {
+                    var pkgCmdSet = VSConstants.VSStd2K;
+                    ErrorHandler.ThrowOnFailure(rw.Exec(ref pkgCmdSet, (uint)VSConstants.VSStd2KCmdID.RETURN, 0, IntPtr.Zero, IntPtr.Zero));
+                } else if (res) {
+                    rw.Submit();
+                } else {
+                    var pkgCmdSet = Microsoft.VisualStudio.Repl.Guids.guidReplWindowCmdSet;
+                    ErrorHandler.ThrowOnFailure(rw.Exec(ref pkgCmdSet, (uint)Microsoft.VisualStudio.Repl.PkgCmdIDList.cmdidBreakLine, 0, IntPtr.Zero, IntPtr.Zero));
+                }
                 return res;
             });
+#endif
 
             if (wait && canExecute) {
-                // A series of quick checks for scenarios where we don't get
-                // notifications via ReadyForInput.
-                for (int checks = 10; checks > 0; --checks) {
-                    if (_replWindowInfo.ReadyForInput.WaitOne(TimeSpan.FromSeconds(0.1))) {
-                        return;
-                    }
-                    if (_window.CaretInStandardInputRegion) {
-                        return;
-                    }
-                }
                 Assert.IsTrue(
-                    _replWindowInfo.ReadyForInput.WaitOne(TimeSpan.FromSeconds(9.0)),
+                    WaitForReadyForInput(TimeSpan.FromSeconds(10)),
                     "Timed out waiting for submitted code to execute: " + line
                 );
             }
         }
 
+        private bool WaitForReadyForInput(TimeSpan timeout) {
+            // A series of quick checks for scenarios where we don't get
+            // notifications via ReadyForInput.
+            for (int checks = 10; checks > 0; --checks) {
+                if (_replWindowInfo.ReadyForInput.WaitOne(TimeSpan.FromSeconds(0.1))) {
+                    return true;
+                }
+                if (IsCaretInStandardInputRegion) {
+                    return true;
+                }
+            }
+            return _replWindowInfo.ReadyForInput.WaitOne(timeout.Subtract(TimeSpan.FromSeconds(1)));

```

Comment:
we don't really know that we waited the full second so we might be shorting the caller up to .9 seconds on this timeout if I understand?  Not a big deal really, but worth mentioning.